angr/engines/vex/light/light.py

```diff
@@ -300,8 +312,42 @@ def _perform_vex_stmt_LoadG(self, addr, alt, guard, dst, cvt, end):
             'ILGop_8Sto32':    ('Ity_I8', 'Iop_8Sto32')     # 8 bit load, S-widen to 32 */
         }
 
+        # Because of how VEX's ARM lifter works, we may introduce non-existent register loads.
+        # Here is an example:
+        #
+        # .text:0800408C ITTTT MI
+        # .text:0800408E LDRMI   R2, =0x40020004
+        # .text:08004090 LDRMI   R3
+        #
+        # 116 | ------ IMark(0x800408e, 2, 1) ------
+        # 117 | t247 = Or32(t225,0x00000040)
+        # 118 | t254 = armg_calculate_condition(t247,t227,t229,t231):Ity_I32
+        # 119 | t262 = GET:I32(r2)
+        # 120 | t263 = CmpNE32(t254,0x00000000)
+        # 121 | t66 = if (t263) ILGop_Ident32(LDle(0x080040bc)) else t262
+        # 122 | PUT(r2) = t66
+        # 123 | PUT(pc) = 0x08004091
+        # 124 | ------ IMark(0x8004090, 2, 1) ------
+        # 125 | t280 = t263
+        # 126 | t73 = if (t280) ILGop_Ident32(LDle(t66)) else t222
+        #
+        # t280 == t263 == the condition inside t66. Now t66 looks like this:
+        #   <BV32 cond then 0x40020004 else reg_r2_861_32{UNINITIALIZED}>. since t280 is guarding the load from t66,
+        # if the load from t66 is not aware of the condition that t280 is True, we will end up reading from r2_861_32,
+        # which is not what the original instruction intended.
+        # Therefore, the load from t66 should be aware of the condition that t280 is True. Or even better, don't
+        # perform the read if the condition is evaluated to False.
+        # We can add another optimization: Let this condition be cond. When cond can be evaluated to either True or
+        # False, we don't want to perform the read when the cond is the guard (which is a relatively cheap check) and
+        # is False. When the cond is True, we perform the read with only the intended address (instead of the entire
+        # guarded address). This way we get rid of the redundant load that should have existed in the first place.
+
         ty, cvt_op = cvt_properties[cvt]
-        load_result = self._perform_vex_stmt_LoadG_load(addr, ty, end)
+        if claripy.is_false(guard[0]):
```

Comment:
This breaks several abstractions! 1) the light engine is not supposed to know what claripy is, 2) by unpacking guard this way you are assuming that the actions mixin is present and no other mixins are messing with the data layer.